### PR TITLE
Create rightscale deployment Destroy actor

### PR DESCRIPTION
**Pull request by siminm**
```
$ nosetests kingpin/actors/rightscale/test/integration_deployment.py --include=integration
....
----------------------------------------------------------------------
Ran 4 tests in 5.862s

OK
```

### CODE REVIEW

File: kingpin/actors/rightscale/deployment.py [hunk 1]
```diff
@@ -37,6 +37,15 @@ class DeploymentBaseActor(base.RightScaleBaseActor):
 
     """Abstract Deployment Actor that provides some utility methods."""
 
+    @gen.coroutine
```
Comment (diranged) [inline kingpin/actors/rightscale/deployment.py]: retry decorator?

File: kingpin/actors/rightscale/deployment.py [hunk 2]
```diff
@@ -98,13 +107,43 @@ def _execute(self):
             self.log.info('Would create a deployment %s' % self.option('name'))
             self.log.debug('Deployment params: %s' % params)
             raise gen.Return()
-        self.log.info('Deployment params: %s' % params)
 
         try:
             yield self._client.create_resource(
```
Comment (diranged) [inline kingpin/actors/rightscale/deployment.py]: Is this wrapped in a retry decorator inside the API class?

File: kingpin/actors/rightscale/deployment.py [hunk 3]
```diff
@@ -98,13 +107,43 @@ def _execute(self):
...
+    """
+
+    all_options = {
+        'name': (str, REQUIRED, 'The name of the deployment to be deleted.'),
+    }
+
+    @gen.coroutine
+    def _execute(self):
+
+        dep = yield self._find_deployment(self.option('name'))
```
Comment (diranged) [inline kingpin/actors/rightscale/deployment.py]: What is raised if this deployment doesn't exist?

File: kingpin/actors/rightscale/test/integration_deployment.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    integration = True
+
+    def setUp(self, *args, **kwargs):
+        super(IntegrationDeployment, self).setUp(*args, **kwargs)
+        self.deployment_name = 'Kingpin Deployment Integration Test'
+
+    @attr('integration', 'dry')
+    @testing.gen_test()
+    def integration_01a_create_deployment_dry(self):
+        actor = deployment.Create('Integ. Test',
```
Comment (diranged) [inline kingpin/actors/rightscale/test/integration_deployment.py]: Make the deployment name more dynamic -- so its unique per test. For example, look at the way we name the ServerArrays in those integration tests.

File: kingpin/actors/rightscale/deployment.py [hunk 5]
```diff
@@ -98,13 +107,43 @@ def _execute(self):
             self.log.info('Would create a deployment %s' % self.option('name'))
             self.log.debug('Deployment params: %s' % params)
             raise gen.Return()
-        self.log.info('Deployment params: %s' % params)
 
         try:
             yield self._client.create_resource(
                 self._client._client.deployments, params)
         except Exception as e:
```
Comment (diranged) [inline kingpin/actors/rightscale/deployment.py]: Wait -- I missed this before. A blanket Exception catch? No no. Please fix.

File: kingpin/actors/rightscale/deployment.py (hunk 1 above)
Comment (siminm) [inline kingpin/actors/rightscale/deployment.py]: this really just calls `find_by_name_nad_keys()` the api module has a retry decorator there

File: kingpin/actors/rightscale/deployment.py (hunk 2 above)
Comment (siminm) [inline kingpin/actors/rightscale/deployment.py]: yes.

File: kingpin/actors/rightscale/deployment.py (hunk 5 above)
Comment (siminm) [inline kingpin/actors/rightscale/deployment.py]: I'd have to catch `requests.http.exceptions` and would import `requests` just for that... that's ok?

Any way to configure `rightscale.settings.RETRYING_SETTINGS` similar to our tornado retry? To raise a specified exception for a matched string?

Comment (nextdoor-ops-ro) [inline kingpin/actors/rightscale/deployment.py]: Go ahead and import the requests library,... its getting loaded up anyways.
:)

Matt Wise
Sr. Systems Architect
Nextdoor.com

**Comment by siminm** [comment]
Addressed exception raising/handling by reworking that logic. Explicitly searching for the deployment before destroying it. This helps with a dry run also, and no need to import `requests` just to catch the exception

File: kingpin/actors/rightscale/deployment.py [hunk 6]
```diff
@@ -98,13 +112,43 @@ def _execute(self):
...
+                'Deployment "%s" does not exist.' % self.option('name'))
+
+        info = (yield self._client.show(dep.self)).soul
+
+        if self._dry:
+            self.log.info('Would delete deployment %s' % info['name'])
+            raise gen.Return()
+
+        self.log.info('Deleting deployment %s' % info['name'])
+        yield self._client.destroy_resource(dep)
```
Comment (diranged) [inline kingpin/actors/rightscale/deployment.py]: Something just occurred to me as I was reading this -- what happens if we try to delete a deployment with servers in it? Is an error returned back to us? Can you try that out real quick?

**Comment by diranged** [comment]
:shipit: ... one quick question in there, but it can be dealt with separately if you want.

**Comment by siminm** [comment]
Hrm, yes it raises RS's catch-all `422` :( Since the destroy actor is there only for completion of integration tests I'd rather deal with it later

**Pull request merged**